### PR TITLE
core/bloombits, eth/filter: transformed bloom bitmap based log search

### PR DESCRIPTION
**Pull request by zsfelfoldi**
This PR is based on https://github.com/ethereum/go-ethereum/pull/14522 and supersedes https://github.com/ethereum/go-ethereum/pull/3749 which was the same thing but based on the old version of the chain processor. Both versions are kept until one of them is merged.

Further parts of the code may be moved into separate PRs to make the review process easier (suggestions are welcome).

This PR optimizes log searching by creating a data structure (BloomBits) that makes it cheaper to retrieve bloom filter data relevant to a specific filter. When searching in a long section of the block history, we are checking three specific bits of each bloom filter per address/topic. In order to do that, currently we read/retrieve a cca. 500 byte block header for each block. The implemented structure optimizes this by a "bitwise 90 degree rotation" of the bloom filters. Blocks are grouped into sections (SectionSize is 4096 blocks at the moment), BloomBits[bitIdx][sectionIdx] is a 4096 bit (512 byte) long bit vector that contains a single bit of each bloom filter from the block range [sectionIdx*SectionSize ... (sectionIdx+1)*SectionSize-1]. (Since bloom filters are usually sparse, a simple data compression makes this structure even more efficient, especially for ODR retrieval.) By reading and binary AND-ing three BloomBits sections, we can filter for an address/topic in 4096 blocks at once ("1" bits in the binary AND result mean bloom matches).

Implementation and design rationale of the matcher logic

- Pipelined structure

The matcher was designed with the needs of both full and light nodes in mind. A simpler architecture would probably be satisfactory for full nodes (where the bit vectors are available in the local database) but the network retrieval bottleneck of light clients justifies a more sophisticated algorithm that tries to minimize the amount of retrieved data and return results as soon as possible. The current implementation is a pipelined structure based on input and output channels (receiving section indexes and sending potential matches). The matcher is built from sub-matchers, one for the addresses and one for each topic group. Since we are interested in matches that each sub-matcher signals as positive, they are daisy-chained in a way that subsequent sub-matchers are only retrieving and matching the bit vectors of sections where the previous matchers have found a potential match. The "1" bits of the output of the last sub-matcher are returned as bloom filter matches.
Sub-matchers use a set of fetchers to retrieve a stream of bit vectors belonging to certain bloom filter bit indexes. Fetchers are also pipelined, receiving a stream of section indexes and sending a stream of bit vectors. As soon as each fetcher returned the bit vector belonging to a certain section index, the sub-matcher performs the necessary binary AND and OR operations and outputs the resulting vector and its belonging section index if the vector is not made of zeroes only.

- Prioritizing and batching requests

Light clients retrieve the bit vectors with merkle proofs, which makes it much more efficient to retrieve batches of vectors (whose merkle proofs share most of their trie nodes) in a single request. Also, it is preferable to prioritize requests based on their section index (regardless of bit index) in order to ensure that matches are found and returned as soon as possible (and in a sequential order). Prioritizing and batching are realized by a common request distributor that receives individual bit index/section index requests from fetchers and keeps an ordered list of section indexes to be requested, grouped by bit index. It does not call any retrieval backend function but it is called by a "server" process (see serveMatcher in filter.go). NextRequest returns the next batch to be requested, retrieved vectors are returned through the Deliver function. This method ensures that the bloombits package should only care about implementing the matching logic. The caller can retain full control over the resources (CPU/disk/network bandwidth) assigned to this task.

### CODE REVIEW

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 738d0f2353505cf3dec13425e5821fd45c43f8e1
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: c13f792d5d1ebc2245a2edfdcb338064818018fa
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

File: core/bloombits/matcher.go [hunk 1]
```diff
@@ -0,0 +1,559 @@
...
+
+// subMatch creates a sub-matcher that filters for a set of addresses or topics, binary OR-s those matches, then
+// binary AND-s the result to the daisy-chain input (processCh) and forwards it to the daisy-chain output.
+// The matches of each address/topic are calculated by fetching the given sections of the three bloom bit indexes belonging to
+// that address/topic, and binary AND-ing those vectors together.
+func (m *Matcher) subMatch(processCh chan partialMatches, idxs []types.BloomIndexList, stop chan struct{}) chan partialMatches {
+	// set up fetchers
+	fetchIdx := make([][3]chan uint64, len(idxs))
+	fetchData := make([][3]chan []byte, len(idxs))
+	for i, idx := range idxs {
```
Comment (Arachnid) [inline core/bloombits/matcher.go]: Can you give 'idxs' and 'ii' more descriptive names?

File: core/bloombits/matcher.go [hunk 2]
```diff
@@ -0,0 +1,559 @@
...
+					close(fetchedCh)
+					for _, ff := range fetchIdx {
+						for _, f := range ff {
+							close(f)
+						}
+					}
+					return
+				}
+
+				for _, ff := range fetchIdx {
```
Comment (Arachnid) [inline core/bloombits/matcher.go]: These variable names are also pretty hard to follow.

File: core/bloombits/matcher.go [hunk 3]
```diff
@@ -0,0 +1,559 @@
...
+			fetchData[i][j] = m.fetchers[ii].fetch(fetchIdx[i][j], m.distCh, stop, &m.wg)
+		}
+	}
+
+	fetchedCh := make(chan partialMatches, channelCap) // entries from processCh are forwarded here after fetches have been initiated
+	resultsCh := make(chan partialMatches, channelCap)
+
+	m.wg.Add(2)
+	// goroutine for starting retrievals
+	go func() {
```
Comment (Arachnid) [inline core/bloombits/matcher.go]: It's unclear to me why there need to be two independent goroutines here, or how they work together. Can you elaborate?

File: core/bloombits/matcher.go [hunk 4]
```diff
@@ -0,0 +1,559 @@
...
+	sectionIdxList []uint64
+}
+
+// distributeRequests receives requests from the fetchers and either queues them
+// or immediately forwards them to one of the waiting NextRequest functions.
+// Requests with a lower section idx are always prioritized.
+func (m *Matcher) distributeRequests(stop chan struct{}) {
+	m.distWg.Add(1)
+	stopDist := make(chan struct{})
+	go func() {
```
Comment (Arachnid) [inline core/bloombits/matcher.go]: Is this really necessary?

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: c13f792d5d1ebc2245a2edfdcb338064818018fa
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 680155f92712a60578260644ff151fc761babc7f
  * Commits must be prefixed with the package(s) they modify

 * Commit: 69049b02777e985be0dd74e7d9a07d95789b3661
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 680155f92712a60578260644ff151fc761babc7f
  * Commits must be prefixed with the package(s) they modify

 * Commit: 69049b02777e985be0dd74e7d9a07d95789b3661
  * Commits must be prefixed with the package(s) they modify

 * Commit: e03a1604ae14c5fb064e143eae19d6ff23e9614e
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 30fe2431930c5b4f4da619f921132b184c3869c3
  * Commits must be prefixed with the package(s) they modify

 * Commit: 771eccad5cc5d386c34875b7e83e6c3f6ac70787
  * Commits must be prefixed with the package(s) they modify

 * Commit: 1d5e142f2ec8ec12901ee8c24a3fbbeb160a335c
  * Commits must be prefixed with the package(s) they modify

 * Commit: 0bc5a7b11ffb0ccb8f3e7977177ba493b8daa216
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 30fe2431930c5b4f4da619f921132b184c3869c3
  * Commits must be prefixed with the package(s) they modify

 * Commit: 771eccad5cc5d386c34875b7e83e6c3f6ac70787
  * Commits must be prefixed with the package(s) they modify

 * Commit: 1d5e142f2ec8ec12901ee8c24a3fbbeb160a335c
  * Commits must be prefixed with the package(s) they modify

 * Commit: 0bc5a7b11ffb0ccb8f3e7977177ba493b8daa216
  * Commits must be prefixed with the package(s) they modify

 * Commit: 7fae8bc809fd83cf4376c8692abc69dc1e65e146
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 30fe2431930c5b4f4da619f921132b184c3869c3
  * Commits must be prefixed with the package(s) they modify

 * Commit: 771eccad5cc5d386c34875b7e83e6c3f6ac70787
  * Commits must be prefixed with the package(s) they modify

 * Commit: 1d5e142f2ec8ec12901ee8c24a3fbbeb160a335c
  * Commits must be prefixed with the package(s) they modify

 * Commit: 0bc5a7b11ffb0ccb8f3e7977177ba493b8daa216
  * Commits must be prefixed with the package(s) they modify

 * Commit: ef1868909d1c5acbd173999e74bd1b4c653ce535
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 30fe2431930c5b4f4da619f921132b184c3869c3
  * Commits must be prefixed with the package(s) they modify

 * Commit: 771eccad5cc5d386c34875b7e83e6c3f6ac70787
  * Commits must be prefixed with the package(s) they modify

 * Commit: 1d5e142f2ec8ec12901ee8c24a3fbbeb160a335c
  * Commits must be prefixed with the package(s) they modify

 * Commit: 0bc5a7b11ffb0ccb8f3e7977177ba493b8daa216
  * Commits must be prefixed with the package(s) they modify

 * Commit: ef1868909d1c5acbd173999e74bd1b4c653ce535
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by karalabe** [comment]
I am now satisfied with this PR, though given that I added heavy reworks to it, I cannot mark it as reviewed. @zsfelfoldi Please double check that the code still makes sense from your perspective. @fjl or @Arachnid Please do a final review.

File: core/bloombits/generator.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	b := &Generator{sections: sections}
+	for i := 0; i < types.BloomBitLength; i++ {
+		b.blooms[i] = make([]byte, sections/8)
+	}
+	return b, nil
+}
+
+// AddBloom takes a single bloom filter and sets the corresponding bit column
+// in memory accordingly.
+func (b *Generator) AddBloom(bloom types.Bloom) error {
```
Comment (Arachnid) [inline core/bloombits/generator.go]: I'm a little concerned that the implied state here (which bit is being set) could lead to something becoming invisibly out of sync. Could this take the bit being set as a parameter and error if it's not the one it expects?

File: core/bloombits/generator.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+	// Make sure we're not adding more bloom filters than our capacity
+	if b.nextBit >= b.sections {
+		return errSectionOutOfBounds
+	}
+	// Rotate the bloom and insert into our collection
+	byteMask := b.nextBit / 8
+	bitMask := byte(1) << byte(7-b.nextBit%8)
+
+	for i := 0; i < types.BloomBitLength; i++ {
+		bloomByteMask := types.BloomByteLength - 1 - i/8
```
Comment (Arachnid) [inline core/bloombits/generator.go]: This is an index, not a mask, isn't it?

File: core/bloombits/matcher.go [hunk 7]
```diff
@@ -0,0 +1,651 @@
...
+// bloomIndexes represents the bit indexes inside the bloom filter that belong
+// to some key.
+type bloomIndexes [3]uint
+
+// calcBloomIndexes returns the bloom filter bit indexes belonging to the given key.
+func calcBloomIndexes(b []byte) bloomIndexes {
+	b = crypto.Keccak256(b)
+
+	var idxs bloomIndexes
+	for i := 0; i < len(idxs); i++ {
```
Comment (Arachnid) [inline core/bloombits/matcher.go]: There's a lot of magic numbers here. Could they be made into constants?

File: core/bloombits/matcher.go [hunk 8]
```diff
@@ -0,0 +1,651 @@
...
+	Bitsets  [][]byte
+}
+
+// Matcher is a pipelined system of schedulers and logic matchers which perform
+// binary AND/OR operations on the bit-streams, creating a stream of potential
+// blocks to inspect for data content.
+type Matcher struct {
+	sectionSize uint64 // Size of the data batches to filter on
+
+	addresses  []bloomIndexes      // Addresses the system is filtering for
```
Comment (Arachnid) [inline core/bloombits/matcher.go]: Wouldn't it be cleaner to just accept a list of byte slices, and leave the address/topic distinction to the caller?

File: core/bloombits/matcher.go [hunk 9]
```diff
@@ -0,0 +1,651 @@
...
+					first = begin
+				}
+				last := sectionStart + m.sectionSize - 1
+				if end < last {
+					last = end
+				}
+				// Iterate over all the blocks in the section and return the matching ones
+				for i := first; i <= last; i++ {
+					// If the bitset is nil, we're a special match-all cornercase
+					if res.bitset == nil {
```
Comment (Arachnid) [inline core/bloombits/matcher.go]: Why the corner case? Why not just return a fully set bitset?

File: core/bloombits/matcher.go [hunk 10]
```diff
@@ -0,0 +1,651 @@
...
+				}
+			}
+		}
+	}()
+	return results
+}
+
+// distributor receives requests from the schedulers and queues them into a set
+// of pending requests, which are assigned to retrievers wanting to fulfil them.
+func (m *Matcher) distributor(dist chan *request, session *MatcherSession) {
```
Comment (Arachnid) [inline core/bloombits/matcher.go]: Does having multiple parallel retrievers help performance? I wonder if it wouldn't be far simpler to just have each filter fetch data as it needs it, with a cache to prevent duplicate reads?

Comment (karalabe) [reply core/bloombits/matcher.go]: I think the complexity is to aid the light client which needs to batch multiple requests together and send them out multiple batches to different light servers. Hence why the whole distribution "mess".

File: core/bloombits/matcher.go (hunk 7 above)
Comment (karalabe) [reply core/bloombits/matcher.go]: These are kind of speced by the yellow paper on how to construct the header bloom filters. I'm unsure if it makes sense to separate them out since there's not much room to reuse them elsewhere. But I'm open to suggestions.

File: core/bloombits/matcher.go (hunk 8 above)
Comment (karalabe) [reply core/bloombits/matcher.go]: API wise within `NewMatcher` I think it might be  nice thing to support the filter by being explicitly catered for that use case. Internally, I'm unsure if we need to retain this distinction. @zsfelfoldi ?

File: core/bloombits/matcher.go (hunk 9 above)
Comment (karalabe) [reply core/bloombits/matcher.go]: Fair enough.

File: core/bloombits/matcher.go (hunk 10 above)
Comment (Arachnid) [reply core/bloombits/matcher.go]: Ah, good point. Objection withdrawn, then.

File: core/bloombits/matcher.go (hunk 8 above)
Comment (Arachnid) [reply core/bloombits/matcher.go]: Even in `NewMatcher`, this seems to me like it shouldn't care about what it's matching on - they should just be opaque hashes to it.

File: core/bloombits/generator.go (hunk 5 above)
Comment (zsfelfoldi) [reply core/bloombits/generator.go]: you're right, done

File: core/bloombits/generator.go (hunk 6 above)
Comment (zsfelfoldi) [reply core/bloombits/generator.go]: it is, and also byteMask is. fixed.

File: core/bloombits/matcher.go (hunk 8 above)
Comment (zsfelfoldi) [reply core/bloombits/matcher.go]: I changed it as @Arachnid suggested in the last commit, now we have to convert addresses and hashes to byte slices in filter.go but maybe it's still nicer this way. If you don't like it, we can drop the last commit.

File: core/bloombits/matcher.go (hunk 9 above)
Comment (zsfelfoldi) [reply core/bloombits/matcher.go]: Because that's what comes from the source: https://github.com/zsfelfoldi/go-ethereum/blob/bloombits2/core/bloombits/matcher.go#L229
I don't want to do unnecessary binary ANDs every time in the first subMatcher so I'm sending nils at the source. But in the corner case there are no subMatchers at all (we are receiving the source at the end) so we have to do some special case handling anyway. We can do it some other way but I'm not sure it's going to be any nicer. But I'm open to suggestions.

Comment (karalabe) [reply core/bloombits/matcher.go]: I've added a commit to always send 0xff-s in the first instance. I don't think doing a binary and on a few bytes will matter given that the data comes from disk or the network.

**Comment by karalabe** [comment]
@Arachnid Issues should be addresses now, PTAL
